xmtp/src/conversations.rs

```diff
@@ -59,18 +71,134 @@ where
         // Adjust for padding and ensure start_time > 0
         std::cmp::max(job.last_run - PADDING_TIME_NS, 0)
     }
+
+    fn create_outbound_payload(
+        &self,
+        session: &mut SessionManager,
+        message: &StoredMessage,
+    ) -> Result<StoredOutboundPayload, ConversationError> {
+        let is_prekey_message = !session.has_received_message();
+
+        let metadata = PadlockMessageSealedMetadata {
+            sender_user_address: self.client.wallet_address(),
+            sender_installation_id: self.client.account.contact().installation_id(),
+            recipient_user_address: session.user_address(),
+            recipient_installation_id: session.installation_id(),
+            is_prekey_message,
+        };
+        // TODO encrypted sealed metadata using sealed sender
+        let sealed_metadata = metadata.encode_to_vec();
+        let message_header = PadlockMessageHeader {
+            sent_ns: message.created_at as u64,
+            sealed_metadata,
+        };
+
+        let payload = PadlockMessagePayload {
+            message_version: PadlockMessagePayloadVersion::One as i32,
+            // TODO sign header - requires exposing a vmac method to sign bytes
+            // rather than string: https://matrix-org.github.io/vodozemac/vodozemac/olm/struct.Account.html#method.sign
+            header_signature: None,
+            convo_id: message.convo_id.clone(),
+            content_bytes: message.content.clone(),
+        };
+        let olm_message = session.encrypt(&payload.encode_to_vec());
+        let ciphertext = match olm_message {
+            OlmMessage::Normal(message) => message.to_bytes(),
+            OlmMessage::PreKey(prekey_message) => prekey_message.to_bytes(),
+        };
+
+        let envelope = PadlockMessageEnvelope {
+            header_bytes: message_header.encode_to_vec(),
+            ciphertext,
+        };
+        Ok(StoredOutboundPayload {
+            created_at_ns: message.created_at,
+            content_topic: build_installation_message_topic(&session.installation_id()),
+            payload: envelope.encode_to_vec(),
+            outbound_payload_state: OutboundPayloadState::Pending as i32,
+        })
+    }
+
+    pub fn process_outbound_message(
+        &self,
+        message: &StoredMessage,
+    ) -> Result<(), ConversationError> {
+        let my_sessions = self
```

Comment:
Having this live outside the transaction does open up the possibility that we could be operating with stale data here (if there were concurrent runs of this process). Ideally this entire function would be a single transaction.

Reply:
I ran into a similar issue with the invite loading last week. 

I have been solving by making methods take in a `conn` as an argument. Was even thinking about making the `Save` and `Store` traits work that way.